scanner/datastore/postgres/manifest_metadata.go

```diff
@@ -0,0 +1,157 @@
+	}
+
+	ctx = zlog.ContextWithValues(ctx, "component", "datastore/postgres/indexerMetadataStore.MigrateManifests")
+
+	// insertMissingManifests inserts missing manifests from the manifest table into manifest_metadata,
+	// and it sets the expiration time to the given expiration.
+	const insertMissingManifests = `
+		INSERT INTO manifest_metadata (manifest_id, expiration)
+		SELECT m.hash, $1
+		FROM manifest m
+		WHERE NOT EXISTS (
+			SELECT FROM manifest_metadata mm WHERE mm.manifest_id = m.hash
+		)
+		RETURNING manifest_id`
+
+	rows, err := i.pool.Query(ctx, insertMissingManifests, expiration.UTC())
+	if err != nil {
+		return nil, err
+	}
+	defer rows.Close()
+	var missingManifests []string
+	for rows.Next() {
+		var manifestID string
+		if err := rows.Scan(&manifestID); err != nil {
+			zlog.Warn(ctx).Err(err).Msg("scanning manifest row")
+			continue
+		}
+		missingManifests = append(missingManifests, manifestID)
+	}
+	if err := rows.Err(); err != nil {
+		return nil, fmt.Errorf("reading manifest metadata rows: %w", err)
+	}
+
+	return missingManifests, nil
+}
+
+func (i *indexerMetadataStore) StoreManifest(ctx context.Context, manifestID string, expiration time.Time) error {
+	ctx = zlog.ContextWithValues(ctx, "component", "datastore/postgres/indexerMetadataStore.StoreManifest")
+
+	// insertManifest inserts the metadata into manifest_metadata, overwriting the previous expiration, if it exists.
+	const insertManifest = `
+		INSERT INTO manifest_metadata (manifest_id, expiration) VALUES
+			($1, $2)
+		ON CONFLICT (manifest_id) DO UPDATE SET expiration = $2`
+
+	_, err := i.pool.Exec(ctx, insertManifest, manifestID, expiration.UTC())
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (i *indexerMetadataStore) ManifestExists(ctx context.Context, manifestID string) (bool, error) {
+	ctx = zlog.ContextWithValues(ctx, "component", "datastore/postgres/indexerMetadataStore.ManifestExists")
+
+	// selectManifest returns 1 if the given manifest exists in the table.
+	// If it does not exist, then no rows will be returned.
+	const selectManifest = `SELECT 1 FROM manifest_metadata WHERE manifest_id = $1`
+
+	row := i.pool.QueryRow(ctx, selectManifest, manifestID)
+	var value int
+	err := row.Scan(&value)
+	if errors.Is(err, pgx.ErrNoRows) {
+		return false, nil
+	}
+	if err != nil {
+		return false, fmt.Errorf("checking if manifest exists: %w", err)
+	}
+
+	// Sanity check the returned value is what is expected.
+	return value == 1, nil
+}
+
+func (i *indexerMetadataStore) GCManifests(ctx context.Context, expiration time.Time) ([]string, error) {
+	ctx = zlog.ContextWithValues(ctx, "component", "datastore/postgres/indexerMetadataStore.GCManifests")
+
+	const deleteManifests = `
+		DELETE FROM manifest_metadata
+		WHERE expiration < $1
+		RETURNING manifest_id`
+
+	// Make this a transaction, as failure to delete the manifest should stop the deletion of its metadata.
+	tx, err := i.pool.Begin(ctx)
```

Comment:
I am wondering if a big transaction to delete multiple manifests in a single shot might become a point of contention in the future, as likely they will be created close to each other in time.

Reply:
> as likely the manifests and metadata will be created close to each other in time.

Can you clarify this part? How would this add to any kind of contention?

Reply:
I can imagine deleting A TON at once could be problematic, but that's why I set the default expiration dates to be some time between 7 days and 30 days from now with the GC checking every 6 hours. Ideally, we have even distribution. Let's say there are 10,000 manifests, then each GC cycle will have to clean up ~110 manifests per cycle (ie about 1.1% of the manifests). I think that's ok, but we should have some kind of metrics to tell us in the future

Reply:
> as likely the manifests and metadata will be created close to each other in time.

Thinking IRs would be likely to be created with a timestamp close to each other; as in when you first install, or when new deployments are done among all services in a cluster, etc. It would mean (hypothetically) that when GC runs likely these IRs would be deleted together, rather than evenly distributed as you described.

Reply:
gotcha.

So upon upgrade to this version, all manifests are given the same timestamp of "now". Originally it was hardcoded to be a random time between 7 and 30 days from "now" (for each entry. As in, the expiration time was chosen for each manifest instead of being the same for all of them). Due to the known issues with 4.5.4, I switched it to just GC them all immediately (the GC will run upon startup, detect all manifests have expired, and clean them)

From here, let's pretend we get a bunch of scan requests at once. We set the manifests' expiration dates to be some time between 7 and 30 days from "now". This is per-manifest. So, we'll have a bunch of manifests which expire within the same 7 - 30 day window, but this is a rather large window.

I realize the math I presented above isn't perfect because this 7 - 30 days window has a bunch of overlap. As in, let's say 100 manifests expired on day 3. They are now eligible to be re-indexed on days 14 - 37, which overlaps with that original window between days 14 - 30. I'm still tempted to think this is ok, but not as good as I was originally thinking. I'll create a ticket to consider looking into the current re-indexing windows

Reply:
I redid the math, and each GC cycle would be responsible for about 1,773 manifests starting on day 30 (assuming 10,000 manifests starting on day 1). I can increase the GC frequency to every 4 hours to improve this to be about 1,182 manifests per cycle. I do fear doing the GC any more frequently may be a bit overkill (keeping in mind not every manifest *needs* to be re-indexed every 7 - 30 days, so overdoing it would be wasteful, and add even more pressure to registries). I think this is ok for now. 10,000 images seems like a lot for a cluster. I'll ask around to see what's a reasonable maximum number of images to consider